src/hotspot/share/gc/g1/g1FullCollector.cpp

```diff
@@ -225,15 +228,17 @@ void G1FullCollector::complete_collection() {
   _heap->print_heap_after_full_collection(scope()->heap_transition());
 }
 
-void G1FullCollector::update_attribute_table(HeapRegion* hr) {
-  if (hr->is_free()) {
```

Comment:
Another item that has been noted in a recent discussion with @albertnetymk is that with this change "Free" regions are also marked as "normal" in the table. It would be better to keep them as "Invalid".

I.e. something like (incorporating @albertnetymk other suggestion):
    if (hr->is_free()) {
      return;
    } else if (hr->is_closed_archive(...) {
      [...]
    } else if (hr->is_pinned() || force_pinned) {
      [...]
    } else {
      [...]
    }

There is no real difference as Free regions should never be referenced anywhere and the code should assert elsewhere. It's still nice to also have Free regions as Invalid in that table though.

Reply:
not exactly. please check previous discussion at https://github.com/openjdk/jdk/pull/2760#discussion_r602144952.

Reply:
Some more investigation and discussion about the BOT handling showed that we need to update the BOT for these Survivor-turned-to-Old regions after all.

The reason is that contrary to what I thought, while BOT can handle queries for object start addresses above the "last known valid entry" mentioned in PR#3356, it only does so slowly, and while updating the BOT itself, not updating that "last known valid entry".
So every time it queries for an object start in such regions, it starts walking from the bottom of that region.

See the call chain `G1BlockOffsetTablePart::block_start` -> `forward_to_block_containing_addr` -> `forward_to_block_containing_addr_slow` where the call to `alloc_block_work` in `g1BlockOffsetTable.cpp:236` only updates local boundary and index (not `_next_offset_threshold` and `_next_offset_index`).

This is a problem for young gcs as this behavior will make them slower than expected. Since it is impossible to make the updates to both `_next_offset_threshold` and `_next_offset_index` atomic, and another issue anyway) I would prefer to penalize full gc (or at least keep old behavior) for that.

Could you add code that walks such full young regions and does the `cross_threshold` thing? It certainly does not need to actually compact these regions.

Thanks,
  Thomas

Reply:
Thanks for discussion! I had the same concern (BOTs of these Survivor-turned-to-Old regions contains no useful info, so might make it very slow when finding block start in these regions), but I was not sure if there is some "lazy" mechanism to fill valid BOTs info for these regions when they are accessed subsequently. I just not have time to do further investigation, now I got the answer from you. 

Sure I will add code to fill valid BOTs info for these young regions by the end of full gc.
I'm not sure if it's OK for me to do this BOTs filling action in a separate issue? As we already had such a long discussion, I think it's might be better for us to initialize another discussion for this specific follow-up issue. Please kindly let me know your thoughts.

Reply:
It is fine with me to fix this separately.

Reply:
Note that if that were fixed, I think PR #3356 would not be needed - because then young region BOTs have the expected contents :) Your call.

Reply:
Thanks Thomas!
Although I'm not sure if pr #3356  is still necessary. But sure, I will add code to fill BOTs for "young" regions as a follow-up fix/enhanacement of this issue, I have just created an bug to track it https://bugs.openjdk.java.net/browse/JDK-8264987.
So after this pr is approved and integrated I will initialize the pr for JDK-8264987.